### PR TITLE
Get Windows network stats directly for Containerd

### LINKED ISSUE #104286: containerd on windows doesn't return network stats
#### What happened:

When curling the endpoint for stats we don't see network stats. 

```
curl http://localhost:8003/api/v1/nodes/3809k8s01000000/proxy/stats/summary
...
 {
   "podRef": {
    "name": "iis-1809-668c5b9c87-nglnq",
    "namespace": "default",
    "uid": "3fe45d67-a25b-4a9c-8895-2fe2ed26164d"
   },
   "startTime": "2021-08-10T00:05:39Z",
   "containers": [
    {
     "name": "iis",
     "startTime": "2021-08-10T00:05:57Z",
     "cpu": {
      "time": "2021-08-11T00:04:19Z",
      "usageNanoCores": 0,
      "usageCoreNanoSeconds": 36281250000
     },
     "memory": {
      "time": "2021-08-11T00:04:19Z",
      "workingSetBytes": 65441792
     },
     "rootfs": {
      "time": "2021-08-11T00:04:10Z",
      "availableBytes": 95124049920,
      "capacityBytes": 136912564224,
      "usedBytes": 104857600,
      "inodesUsed": 0
     },
     "logs": {
      "time": "2021-08-11T00:04:19Z",
      "availableBytes": 95124049920,
      "capacityBytes": 136912564224,
      "usedBytes": 0,
      "inodesUsed": 0
     }
    }
   ],
   "cpu": {
    "time": "2021-08-11T00:04:19Z",
    "usageNanoCores": 0,
    "usageCoreNanoSeconds": 36281250000
   },
   "memory": {
    "time": "2021-08-11T00:04:19Z",
    "availableBytes": 0,
    "usageBytes": 0,
    "workingSetBytes": 65441792,
    "rssBytes": 0,
    "pageFaults": 0,
    "majorPageFaults": 0
   },
   "volume": [
    {
     "time": "2021-08-10T00:05:27Z",
     "availableBytes": 96947462144,
     "capacityBytes": 136912564224,
     "usedBytes": 7182,
     "inodesFree": 0,
     "inodes": 0,
     "inodesUsed": 0,
     "name": "kube-api-access-82bxs"
    }
   ],
   "ephemeral-storage": {
    "time": "2021-08-11T00:04:19Z",
    "availableBytes": 95124049920,
    "capacityBytes": 136912564224,
    "usedBytes": 104857821,
    "inodesUsed": 0
   }
  }
```

#### What you expected to happen:

Network stats to be included when calling the summary.  This works with dockershim for windows.  

#### Anything else we need to know?:

This happens because Windows containerd uses hcsv2 to create containers and we query for the networks stats using hcsv1: 

https://github.com/kubernetes/kubernetes/blob/938d31e273328bc3982e74efd84be83ddd3698ea/pkg/kubelet/stats/cri_stats_provider_windows.go#L75

This will be address eventually with move to CRI only metrics KEP: https://github.com/kubernetes/enhancements/pull/2364/files but is still several releases away.

#### Environment:
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration:
- OS (e.g: `cat /etc/os-release`):
- Kernel (e.g. `uname -a`):
- Install tools:
- Network plugin and version (if this is a network-related bug):
- Others:

/sig windows
/assign

### PR DESCRIPTION
**Pull request by jsturtevant**
#### What type of PR is this?
/kind bug
/kind cleanup

#### What this PR does / why we need it:

This ensures we have Network stats for containerd on Windows by querying for the stats were directly.  By going to network component directly we also minimize the calls the to listing containers (https://github.com/kubernetes/kubernetes/issues/104285)

#### Which issue(s) this PR fixes:

Fixes #104286 #104285

#### Does this PR introduce a user-facing change?

```release-note
Add support for Windows Network stats in Containerd
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by jsturtevant** [comment]
Although, it is slightly awkward due to build constraints in hcsshim (can't import the project when GOOS=linux) I added unit tests to the solution to help make sure I got the logic correct since it was a bit convoluted. You can follow the commits to see the unit tests working against existing solution, then the updates to make the new solution work.  

If preferred I can drop out the unit tests since it adds the awkward interface to the provider but they were helpful to aid in the refactoring.

I also added custom logging (now removed) to verify that the data was the same when the same pods where deployed on the system. The result at the point where it is being used is the same minus small differences in actually network traffic:

Before:
```
kubelet.exe.INFO:2979:I1101 23:05:09.067962    9548 cri_stats_provider.go:437] "jjs adding stats" name="d8da0fa244ff18484cd4d0ff4c0a050fd5b843a14d4135ee409f45e7ac5d1f90_flannel.4096" in=3737835 out=230391                                                                                                                                   
kubelet.exe.INFO:2980:I1101 23:05:09.069377    9548 cri_stats_provider.go:437] "jjs adding stats" name="d8da0fa244ff18484cd4d0ff4c0a050fd5b843a14d4135ee409f45e7ac5d1f90_flannel.4096" in=3737835 out=230391                                                                                                                                   
kubelet.exe.INFO:2981:I1101 23:05:09.069432    9548 cri_stats_provider.go:437] "jjs adding stats" name="d8da0fa244ff18484cd4d0ff4c0a050fd5b843a14d4135ee409f45e7ac5d1f90_flannel.4096" in=3737835 out=230391                                                                                                                                   
kubelet.exe.INFO:2982:I1101 23:05:09.069432    9548 cri_stats_provider.go:437] "jjs adding stats" name="fd3971a07bc0ad5a361673469eeaa7e736cfd2cc4a5d799aa4eff0728ba166b5_flannel.4096" in=960598 out=450471                                                                                                                                    
kubelet.exe.INFO:2983:I1101 23:05:09.070791    9548 cri_stats_provider.go:437] "jjs adding stats" name="a06c56ce67b9d6f2857913c95afe662a285797d4a13ef6cc6fb561959dfa5818_flannel.4096" in=2467971 out=994455                                                                                                                                   
kubelet.exe.INFO:2984:I1101 23:05:09.070791    9548 cri_stats_provider.go:437] "jjs adding stats" name="Ethernet" in=3290739 out=1109478                                                         
```

After:
```
kubelet.exe.INFO:121:I1103 00:15:43.348647    9348 cri_stats_provider.go:438] "jjs adding stats" name="d8da0fa244ff18484cd4d0ff4c0a050fd5b843a14d4135ee409f45e7ac5d1f90_flannel.4096" in=43022381 out=3740326
kubelet.exe.INFO:122:I1103 00:15:43.349201    9348 cri_stats_provider.go:438] "jjs adding stats" name="d8da0fa244ff18484cd4d0ff4c0a050fd5b843a14d4135ee409f45e7ac5d1f90_flannel.4096" in=43022381 out=3740326
kubelet.exe.INFO:123:I1103 00:15:43.349274    9348 cri_stats_provider.go:438] "jjs adding stats" name="d8da0fa244ff18484cd4d0ff4c0a050fd5b843a14d4135ee409f45e7ac5d1f90_flannel.4096" in=43022381 out=3740326
kubelet.exe.INFO:124:I1103 00:15:43.349878    9348 cri_stats_provider.go:438] "jjs adding stats" name="fd3971a07bc0ad5a361673469eeaa7e736cfd2cc4a5d799aa4eff0728ba166b5_flannel.4096" in=6781898 out=4690117
kubelet.exe.INFO:125:I1103 00:15:43.350432    9348 cri_stats_provider.go:438] "jjs adding stats" name="a06c56ce67b9d6f2857913c95afe662a285797d4a13ef6cc6fb561959dfa5818_flannel.4096" in=3237044 out=1365159
kubelet.exe.INFO:126:I1103 00:15:43.351104    9348 cri_stats_provider.go:438] "jjs adding stats" name="Ethernet" in=4161468 out=1509101
```

I added e2e test to make sure these stats are returned so they aren't dropped in the future as well.

Finally I collected results against the kubelet and vmcompute while running the modified e2e test to verify cpu went down when removing the duplicated calls as identified in https://github.com/kubernetes/kubernetes/issues/104285. And you can see a difference (the right half is when the test was run)

before:
image


after:
image

File: pkg/kubelet/stats/cri_stats_provider.go [hunk 1]
```diff
@@ -65,6 +65,9 @@ type criStatsProvider struct {
 	imageService internalapi.ImageManagerService
 	// hostStatsProvider is used to get the status of the host filesystem consumed by pods.
 	hostStatsProvider HostStatsProvider
+	//lint:ignore U1000 We can't import hcsshim due to Build constraints in hcsshim
+	// hcsshimInterface is used by Windows kubelet to gather stats for the for networking
```
Comment (marosset) [inline pkg/kubelet/stats/cri_stats_provider.go:69]:
```suggestion
	// hcsshimInterface is used by kubelet to gather networking stats on Windows
```

File: pkg/kubelet/stats/cri_stats_provider_windows.go [hunk 2]
```diff
@@ -20,104 +20,102 @@ limitations under the License.
...
 // listContainerNetworkStats returns the network stats of all the running containers.
 func (p *criStatsProvider) listContainerNetworkStats() (map[string]*statsapi.NetworkStats, error) {
-	containers, err := hcsshim.GetContainers(hcsshim.ComputeSystemQuery{
-		Types: []string{"Container"},
-	})
+	shim := newHcsShim(p)
+
+	endpoints, err := shim.HNSListEndpointRequest()
 	if err != nil {
+		klog.ErrorS(err, "Failed to fetch current endpoints endpoints")
```
Comment (marosset) [inline pkg/kubelet/stats/cri_stats_provider_windows.go:59]:
```suggestion
		klog.ErrorS(err, "Failed to fetch current endpoints", endpoints)
```
Was this the intended behavior here?

**Comment by marosset** [comment]
I would definitely like @dcantah to review here

**Comment by dcantah** [comment]
@marosset Currently looking :P

File: pkg/kubelet/stats/cri_stats_provider_windows.go (hunk 2 above)
Comment (jsturtevant) [reply pkg/kubelet/stats/cri_stats_provider_windows.go:59]: it was duplicated by mistake, will fix this and the other typo

File: test/e2e/windows/kubelet_stats.go [hunk 3]
```diff
@@ -80,6 +80,11 @@ var _ = SIGDescribe("[Feature:Windows] Kubelet-Stats [Serial]", func() {
 						framework.ExpectEqual(*podStats.CPU.UsageCoreNanoSeconds > 0, true, "Pod stats should not report 0 cpu usage")
 						framework.ExpectEqual(*podStats.Memory.WorkingSetBytes > 0, true, "Pod stats should not report 0 bytes for memory working set ")
 
+						framework.ExpectEqual(podStats.Network != nil, true, "Pod stats should report network stats")
+						framework.ExpectEqual(podStats.Network.Name != "", true, "Pod stats should report network name")
+						framework.ExpectEqual(*podStats.Network.TxBytes > 0, true, "Pod stats should report network Tx stats")
+						framework.ExpectEqual(len(podStats.Network.Interfaces) > 0, true, "Pod Stats should report individual interfaces stats")
+
```
Comment (dcantah) [inline test/e2e/windows/kubelet_stats.go:87]: We're expecting `Network` to be filled in now but there may be a nil deref for lines 84-86 if for some reason this ever changes right?

File: pkg/kubelet/stats/cri_stats_provider_windows_test.go [hunk 4]
```diff
@@ -0,0 +1,424 @@
...
+							RxBytes: toP(2),
+							TxBytes: toP(20),
+						},
+					},
+				},
+			},
+			wantErr: false,
+		},
+		{
+			name: "multiple endpoints per container",
```
Comment (dcantah) [inline pkg/kubelet/stats/cri_stats_provider_windows_test.go:326]: Ignorance incoming: Are there Windows CNI's that allocate multiple endpoints per pod?

File: pkg/kubelet/stats/cri_stats_provider.go [hunk 5]
```diff
@@ -65,6 +65,9 @@ type criStatsProvider struct {
...
+	hcsshimInterface interface{}
```
Comment (dcantah) [inline pkg/kubelet/stats/cri_stats_provider.go:70]: I'm not too fond of using the hcsshim name explicitly here (or any of the other uses in the other files). Would windowsStatsInterface be a suitable replacement?

File: pkg/kubelet/stats/cri_stats_provider_windows_test.go (hunk 4 above)
Comment (jsturtevant) [reply pkg/kubelet/stats/cri_stats_provider_windows_test.go:326]: I am not aware of any but the previous implementation did handle the scenario which is why added this test case

File: pkg/kubelet/stats/cri_stats_provider.go (hunk 5 above)
Comment (jsturtevant) [reply pkg/kubelet/stats/cri_stats_provider.go:70]: Works for me, I don't really like the way this is interface is used but couldn't find another way to do this given the build constraints. Would be interested in feedback if it is worth this for having the unit tests.  The unit tests were helpful for me to get the logic right.

File: test/e2e/windows/kubelet_stats.go (hunk 3 above)
Comment (jsturtevant) [reply test/e2e/windows/kubelet_stats.go:87]: There is a `nil` check on line 83 where the test would stop processing and report `network` was found to be `nil`

Comment (dcantah) [reply test/e2e/windows/kubelet_stats.go:87]: Aha, didn't know what the behavior for ExpectEqual was (e.g. immediate failure). Feel free to resolve!

File: pkg/kubelet/stats/cri_stats_provider.go (hunk 5 above)
Comment (jsturtevant) [reply pkg/kubelet/stats/cri_stats_provider.go]: updated

File: pkg/kubelet/stats/cri_stats_provider_windows.go (hunk 2 above)
Comment (jsturtevant) [reply pkg/kubelet/stats/cri_stats_provider_windows.go]: updated

File: pkg/kubelet/stats/cri_stats_provider.go (hunk 1 above)
Comment (jsturtevant) [reply pkg/kubelet/stats/cri_stats_provider.go]: updated

File: pkg/kubelet/stats/cri_stats_provider_windows.go [hunk 6]
```diff
@@ -20,104 +20,102 @@ limitations under the License.
...
 	"github.com/Microsoft/hcsshim"
 
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
-	"k8s.io/apimachinery/pkg/util/sets"
-	"k8s.io/klog/v2"
 	statsapi "k8s.io/kubelet/pkg/apis/stats/v1alpha1"
 )
 
+// hcsShimInterface creates an interface that allows for testing the logic without needing to create a container
+type hcsShimInterface interface {
```
Comment (dcantah) [inline pkg/kubelet/stats/cri_stats_provider_windows.go:34]: I'd prefer the windowsNetworkStatsInterface for this and any mentions of hcsshim also

Comment (jsturtevant) [reply pkg/kubelet/stats/cri_stats_provider_windows.go]: updated

File: pkg/kubelet/stats/cri_stats_provider.go (hunk 5 above)
Comment (jsturtevant) [reply pkg/kubelet/stats/cri_stats_provider.go]: @SergeyKanzhelev do you have any thoughts on if this is an ok approach to so unit tests can be enabled or should we drop this (and the unit tests).

File: pkg/kubelet/stats/cri_stats_provider_windows_test.go (hunk 4 above)
Comment (dcantah) [reply pkg/kubelet/stats/cri_stats_provider_windows_test.go:326]: Gotcha, feel free to resolve

File: pkg/kubelet/stats/cri_stats_provider_windows.go [hunk 7]
```diff
@@ -20,104 +20,101 @@ limitations under the License.
...
-	"k8s.io/apimachinery/pkg/util/sets"
-	"k8s.io/klog/v2"
 	statsapi "k8s.io/kubelet/pkg/apis/stats/v1alpha1"
 )
 
+// windowsNetworkStatsProvider creates an interface that allows for testing the logic without needing to create a container
+type windowsNetworkStatsProvider interface {
+	HNSListEndpointRequest() ([]hcsshim.HNSEndpoint, error)
+	GetHNSEndpointStats(endpointName string) (*hcsshim.HNSEndpointStats, error)
+	Now() time.Time
```
Comment (dashpole) [inline pkg/kubelet/stats/cri_stats_provider_windows.go:36]: Not sure Now() belongs in the interface.  Adding a k8s.io/utils/clock.Clock to the cri stats provider seems better.

**Review by dashpole — COMMENTED** [review]
/approve
/hold
Adding hold to give you a chance to address comments.  Remove when you are satisfied

**Comment by SergeyKanzhelev** [comment]
/cc @bobbypage @qiutongs 

FYI

File: pkg/kubelet/stats/cri_stats_provider_windows.go (hunk 7 above)
Comment (jsturtevant) [reply pkg/kubelet/stats/cri_stats_provider_windows.go]: makes sense, i've made this adjustment

**Comment by ehashman** [comment]
/hold cancel
@dashpole's comment is addressed

**Comment by bobbypage** [comment]
Thank you @jsturtevant for adding this. I wanted to point out the [KEP](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2371-cri-pod-container-stats) that me and @haircommander are working on to move all of the stats from pod / container stats from the kubelet to the underlying container runtime (this includes network stats). So long term, it would be great to get these stats implemented directly in containerd CRI API (and get support on windows!). Happy to chat more about this offline, will ping you on slack :)

**Comment by jsturtevant** [comment]
yup been following the effort!  lets chat in slack

**Pull request merged**